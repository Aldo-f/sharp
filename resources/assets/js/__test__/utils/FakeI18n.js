export default {
    install() {
        window.i18n = {};
    }
}