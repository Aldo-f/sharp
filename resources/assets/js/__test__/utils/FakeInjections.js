import Vue from 'vue';

const injectedComponents = ['$field', '$form'];

export default {
    provide() {
        return injectedComponents.reduce((injections, prop) => {
            injections[prop] = new Vue();
            return injections;
        }, {});
    }
}