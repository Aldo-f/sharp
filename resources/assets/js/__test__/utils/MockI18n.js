import Localization ,* as localizationFn from '../../mixins/Localization';

const mockLangImplementation = localeKey => `{{ ${localeKey} }}`;

export default {
    install() {
        window.i18n = {};
    },

    mockLangFunction() {
        localizationFn.lang = jest.fn(mockLangImplementation);
        Localization.methods.l = jest.fn(mockLangImplementation);
    }
}