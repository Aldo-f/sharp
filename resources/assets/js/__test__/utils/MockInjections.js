import Vue from 'vue';

const injectedComponents = {
    $field:{},
    $form:{
        data: () => ({
            errors: {}
        })
    },
    $tab:{}
};

export default {
    provide() {
        return Object.keys(injectedComponents).reduce((injections, compName) => {
            let Comp = Vue.extend(injectedComponents[compName]);
            injections[compName] = new Comp();
            return injections;
        }, {});
    }
}