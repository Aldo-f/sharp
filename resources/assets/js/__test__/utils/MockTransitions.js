export default {
    install() {
        window.getComputedStyle = jest.fn(() => ({
            animationDelay:'',
            animationDuration:'',
            transitionDelay:'',
            transitionDuration:''
        }));
    }
}