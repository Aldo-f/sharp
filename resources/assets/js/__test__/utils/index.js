export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC } from './mockSFC';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}