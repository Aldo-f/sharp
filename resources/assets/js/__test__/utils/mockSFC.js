export function mockSFC (sfc, extend) {
    Object.assign(sfc, {
        mixins: [],
        'extends': extend,
        template: null,
        render() { return this._v(`__MOCKED_SFC_${sfc.name}__\n`) },
        watch: {},
        beforeCreate: ()=>{},
        created: ()=>{},
        beforeMount: ()=>{},
        mounted: ()=>{},
        beforeUpdate: ()=>{},
        updated: ()=>{},
        beforeDestroy: ()=>{},
        destroyed: ()=>{},
        data: ()=>({}),
        provide: {},
        inject: [],
    })
}