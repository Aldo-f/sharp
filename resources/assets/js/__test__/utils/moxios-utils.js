import moxios from 'moxios';

export function nextRequestFulfilled(config) {
    return new Promise(resolve => {
        moxios.wait(async () => {
            let request = moxios.requests.mostRecent();
            let response = await request.respondWith(config);
            resolve(response); // moxios Response object
        })
    });
}
