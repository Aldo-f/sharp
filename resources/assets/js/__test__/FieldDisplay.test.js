import Vue from 'vue';
import * as fieldContainerModule from '../components/form/FieldContainer.vue';

import { MockInjections, mockSFC } from "./utils/index";

import FieldDisplay from '../components/form/field-display/FieldDisplay';
import * as conditions from '../components/form/field-display/conditions';

describe('field-display', () => {
    Vue.component('sharp-field-display',FieldDisplay);

    let computeSelectCondition = ()=>{}, computeCondition = ()=>{};

    beforeEach(() => {
        mockSFC(fieldContainerModule.default);

        document.body.innerHTML = `
            <div id="app">
                <sharp-field-display
                    field-key="title"
                    :context-data="contextData"
                    :context-fields="contextFields"
                    :update-visibility="updateVisibility"
                    config-identifier="title"
                    error-identifier="title"
                >
                </sharp-field-display> 
            </div>
        `;

        computeSelectCondition = conditions.helpers.computeSelectCondition = jest.fn(conditions.helpers.computeSelectCondition);
        computeCondition = conditions.computeCondition = jest.fn(conditions.computeCondition);
    });

    it('can mount field display', async () => {
        await createVm({
            propsData: {
                contextData: {
                    title: null
                },
                contextFields: {
                    title: { type: 'text' }
                }
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    describe('conditional diplay', ()=>{

        it('check', async () => {
            let vm = await createVm({
                propsData: {
                    contextData: {
                        title: null,
                        check: true
                    },
                    contextFields: {
                        title: {
                            type: 'text',
                            conditionalDisplay: {
                                operator: 'or',
                                fields: [{ key:'check', values: true }]
                            }
                        },
                        check: { type: 'check' }
                    }
                }
            });


            expect(vm.$children).toHaveLength(1);
            expect(computeCondition)
                .toHaveBeenCalledWith(vm.contextFields, vm.contextData, vm.contextFields.title.conditionalDisplay);

            Vue.set(vm.contextData,'check',false);
            await Vue.nextTick();

            expect(vm.$children).toHaveLength(0);
        });

        it('multiple select (unique values)', async () => {
            let vm = await createVm({
                propsData: {
                    contextData: {
                        title: null,
                        select: [ 2, 3 ]
                    },
                    contextFields: {
                        title: {
                            type: 'text',
                            conditionalDisplay: {
                                operator: 'or',
                                fields: [
                                    { key:'select', values: 2 }
                                ]
                            }
                        },
                        select: {
                            type: 'select',
                            multiple: true
                        }
                    }
                }
            });

            expect(computeSelectCondition).toHaveBeenCalledWith({ condValues: 2, fieldValue: [2,3], isSingleSelect: false });
            expect(vm.$children).toHaveLength(1);

            Vue.set(vm.contextData, 'select', [ 3 ]);
            await Vue.nextTick();

            expect(computeSelectCondition).toHaveBeenCalledWith({ condValues: 2, fieldValue: [3], isSingleSelect: false });
            expect(vm.$children).toHaveLength(0);
        });

        it('multiple select (multiple values)', async () => {
            let vm = await createVm({
                propsData: {
                    contextData: {
                        title: null,
                        select: [ 2, 3 ]
                    },
                    contextFields: {
                        title: {
                            type: 'text',
                            conditionalDisplay: {
                                operator: 'or',
                                fields: [
                                    { key:'select', values: [ 2, 6 ] }
                                ]
                            }
                        },
                        select: {
                            type: 'select',
                            multiple: true
                        }
                    }
                }
            });

            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: [2,6], fieldValue: [2,3], isSingleSelect: false });
            expect(vm.$children).toHaveLength(1);

            Vue.set(vm.contextData, 'select', [ 4, 5 ]);
            await Vue.nextTick();


            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: [2,6], fieldValue: [4,5], isSingleSelect: false });
            expect(vm.$children).toHaveLength(0);
        });

        it('single select (unique values)', async () => {
            let vm = await createVm({
                propsData: {
                    contextData: {
                        title: null,
                        select: 4
                    },
                    contextFields: {
                        title: {
                            type: 'text',
                            conditionalDisplay: {
                                operator: 'or',
                                fields: [
                                    { key:'select', values: 4 }
                                ]
                            }
                        },
                        select: {
                            type: 'select'
                        }
                    }
                }
            });

            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: 4, fieldValue: 4, isSingleSelect: true });
            expect(vm.$children).toHaveLength(1);

            Vue.set(vm.contextData, 'select', 5);
            await Vue.nextTick();

            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: 4, fieldValue: 5, isSingleSelect: true });
            expect(vm.$children).toHaveLength(0);
        });

        it('single select (multiple values)', async () => {
            let vm = await createVm({
                propsData: {
                    contextData: {
                        title: null,
                        select: 4
                    },
                    contextFields: {
                        title: {
                            type: 'text',
                            conditionalDisplay: {
                                operator: 'or',
                                fields: [
                                    { key:'select', values: [4, 8] }
                                ]
                            }
                        },
                        select: {
                            type: 'select'
                        }
                    }
                }
            });

            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: [4,8], fieldValue: 4, isSingleSelect: true });
            expect(vm.$children).toHaveLength(1);

            Vue.set(vm.contextData, 'select', 3);
            await Vue.nextTick();

            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: [4,8], fieldValue: 3, isSingleSelect: true });
            expect(vm.$children).toHaveLength(0);
        });

        it('multiple select (negative)', async () => {
            let vm = await createVm({
                propsData: {
                    contextData: {
                        title: null,
                        select: [ 4, 6 ]
                    },
                    contextFields: {
                        title: {
                            type: 'text',
                            conditionalDisplay: {
                                operator: 'or',
                                fields: [
                                    { key:'select', values: '!5' }
                                ]
                            }
                        },
                        select: {
                            type: 'select',
                            multiple: true
                        }
                    }
                }
            });

            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: '!5', fieldValue: [4,6], isSingleSelect: false });
            expect(vm.$children).toHaveLength(1);

            Vue.set(vm.contextData, 'select', [ 5, 7 ]);
            await Vue.nextTick();

            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: '!5', fieldValue: [5,7], isSingleSelect: false });
            expect(vm.$children).toHaveLength(0);
        });

        it('single select (negative)', async () => {
            let vm = await createVm({
                propsData: {
                    contextData: {
                        title: null,
                        select: 3
                    },
                    contextFields: {
                        title: {
                            type: 'text',
                            conditionalDisplay: {
                                operator: 'or',
                                fields: [
                                    { key:'select', values: '!6' }
                                ]
                            }
                        },
                        select: {
                            type: 'select'
                        }
                    }
                }
            });

            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: '!6', fieldValue: 3, isSingleSelect: true });
            expect(vm.$children).toHaveLength(1);

            Vue.set(vm.contextData, 'select', 6);
            await Vue.nextTick();

            expect(computeSelectCondition).toHaveBeenLastCalledWith({ condValues: '!6', fieldValue: 6, isSingleSelect: true });
            expect(vm.$children).toHaveLength(0);
        });

        it('or operator', async () => {
            let vm = await createVm({
                propsData: {
                    contextData: {
                        title: null,
                        check1: true,
                        check2: true,
                    },
                    contextFields: {
                        title: {
                            type: 'text',
                            conditionalDisplay: {
                                operator: 'or',
                                fields: [
                                    { key:'check1', values: true },
                                    { key:'check2', values: true }
                                ]
                            }
                        },
                        check1: { type: 'check' },
                        check2: { type: 'check' }
                    }
                }
            });

            expect(vm.$children).toHaveLength(1);

            Vue.set(vm.contextData,'check1',false);
            await Vue.nextTick();

            expect(vm.$children).toHaveLength(1);

            Vue.set(vm.contextData,'check2',false);
            await Vue.nextTick();

            expect(vm.$children).toHaveLength(0);
        });

        it('and operator', async () => {
            let vm = await createVm({
                propsData: {
                    contextData: {
                        title: null,
                        check1: true,
                        check2: true,
                    },
                    contextFields: {
                        title: {
                            type: 'text',
                            conditionalDisplay: {
                                operator: 'and',
                                fields: [
                                    { key:'check1', values: true },
                                    { key:'check2', values: true }
                                ]
                            }
                        },
                        check1: { type: 'check' },
                        check2: { type: 'check' }
                    }
                }
            });

            expect(vm.$children).toHaveLength(1);

            Vue.set(vm.contextData,'check1',false);
            await Vue.nextTick();
            
            expect(vm.$children).toHaveLength(0);
        });
    });
});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [MockInjections, customOptions],

        props:['contextData', 'contextFields'],
        
        'extends': {
            data:()=>({
                value: null
            }),
            methods:{
                updateVisibility: ()=>{}
            }
        }
    });

    await Vue.nextTick();

    return vm;
}