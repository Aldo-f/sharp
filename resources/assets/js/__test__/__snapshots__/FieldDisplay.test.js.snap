// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[check] display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[check] don't display 1`] = `
<div id="app">
    <!---->
</div>
`;

exports[`[multiple select#multiple values] display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[multiple select#multiple values] don't display 1`] = `
<div id="app">
    <!---->
</div>
`;

exports[`[multiple select#negative] display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[multiple select#negative] don't display 1`] = `
<div id="app">
    <!---->
</div>
`;

exports[`[multiple select#unique values] display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[multiple select#unique values] don't display 1`] = `
<div id="app">
    <!---->
</div>
`;

exports[`[operator AND] display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[operator AND] don't display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[operator OR] display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[operator OR] display 2`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[operator OR] don't display 1`] = `
<div id="app">
    <!---->
</div>
`;

exports[`[single select#multiple values] display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[single select#multiple values] don't display 1`] = `
<div id="app">
    <!---->
</div>
`;

exports[`[single select#negative] display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[single select#negative] don't display 1`] = `
<div id="app">
    <!---->
</div>
`;

exports[`[single select#unique values] display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;

exports[`[single select#unique values] don't display 1`] = `
<div id="app">
    <!---->
</div>
`;

exports[`field-display can mount field display 1`] = `
<div id="app">
    __MOCKED_SFC_SharpFieldContainer__
</div>
`;
